network/p2p/connection.go

```diff
@@ -82,32 +87,166 @@ func (conn *connection) send(message *transport.NetworkMessage) error {
 	return nil
 }
 
-// sendMessages (blocking) reads messages from its outMessages channel and sends them to the peer until the channel is closed.
-func (conn connection) sendMessages() {
-	for message := range conn.outMessages {
-		if conn.messenger.Send(message) != nil {
-			log.Logger().Warnf("Unable to send message to peer (peer=%s)", conn.Peer)
+func (conn *managedConnection) exchange(receivedMessages messageQueue) {
+	conn.mux.Lock()
+	out := conn.outMessages
+	in := conn.receiveMessages()
+	conn.mux.Unlock()
+	for {
+		select {
+		case message := <-out:
+			if message == nil {
+				// Connection is closing
+				return
+			}
+			if conn.messenger.Send(message) != nil {
+				log.Logger().Warnf("Unable to send message to peer (peer=%s)", conn.Peer)
+			}
+		case message := <-in:
+			if message == nil {
+				// Connection is closing
+				return
+			}
+			receivedMessages.c <- *message
+		case <-conn.closer:
+			log.Logger().Trace("close() is called")
+			return
 		}
 	}
 }
 
-// receiveMessages (blocking) reads messages from the peer until it disconnects or the network is stopped.
-func (conn *connection) receiveMessages(receivedMsgQueue messageQueue) {
-	for {
-		msg, recvErr := conn.messenger.Recv()
-		if recvErr != nil {
-			if recvErr == io.EOF {
-				log.Logger().Infof("Peer closed connection (peer-id=%s)", conn.ID)
-			} else {
-				log.Logger().Warnf("Peer connection error (peer-id=%s): %v", conn.ID, recvErr)
-			}
-			break
+func (conn *managedConnection) close() {
```

Comment:
Does close need to do anything with `conn.messenger`?

Reply:
Not necessarily (because the connection will get GC'd anyways), but it might be more clean (`nil` all fields when `close()` is called) and makes it impossible to accidentally receive/send messages after it has been closed. Will add.